File: src/sorting/MergeSort.java
import java.util.Arrays;
public class MergeSort {
    public static void main(String[] args) {
        long[] inputList = new long[args.length];
        for (int i = 0; i < args.length; i++) {
            inputList[i] = Long.parseLong(args[i]);
        }
        long [] outputList = splitList(inputList);
        IO.println(Arrays.toString(outputList));
    }

    public static long[] splitList(long[] list) {
        if (list.length < 2) {
            return list;
        }

        int mid = list.length / 2;
        long[] left = new long[mid];
        long[] right = new long[list.length - mid];
        for (int i = 0; i < list.length; i++) {
            if (i < mid) {
                left[i] = list[i];
            }
            else {
                right[i - mid] = list[i];
            }
        }
        long[] L = splitList(left);
        long[] R = splitList(right);
        return mergeLists(L, R);
    }

    public static long[] mergeLists(long[] left, long[] right) {
        long[] result = new long[left.length + right.length];
        int resultIndex = 0;
        int i = 0;
        int j = 0;

        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]){
                result[resultIndex] = left[i];
                i++;
            } else {
                result[resultIndex] = right[j];
                j++;
            }
            resultIndex++;
        }

        while (i < left.length) {
            result[resultIndex] = left[i];
            i++;
            resultIndex++;
        }
                
        while (j < right.length) {
            result[resultIndex] = right[j];
            j++;
            resultIndex++;
        }
        return result;
    }
}

File: src/sorting/merge_sort.js
function sortArray(nums: number[]): number[] {
    if (nums === undefined){
        return [];
    }
    if (nums.length < 2) {
        return nums;
    }
    const mid = Math.floor(nums.length / 2);
    const left = sortArray(nums.slice(0,mid));
    const right = sortArray(nums.slice(mid, nums.length));
    return merge(left, right);
};

function merge(left: number[], right: number[]): number[] {
    const result = [];
    let i = 0;
    let j = 0;
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }
    while (i < left.length) {
        result.push(left[i]);
        i++;
    }
    while (j < right.length) {
        result.push(right[j]);
        j++;
    }
    return result;
}
